src/worker.ts

```diff
@@ -228,73 +221,84 @@ class Reactor {
           }
           break;
 
+        // a common invalid state:
+        // case 0:
+        //   console.log({retval, input: reader.input, reader })
         default:
           throw new Error(
-            `invalid section type="${sectionType}"; please open an issue (https://github.com/extism/js-sdk/issues/new?title=shared+array+buffer+bad+section+type+${sectionType}&labels=bug)`,
+            `invalid section type="${sectionType}" at position ${reader.position}; please open an issue (https://github.com/extism/js-sdk/issues/new?title=shared+array+buffer+bad+section+type+${sectionType}&labels=bug)`,
           );
-          break;
       }
     } while (1);
   }
 }
 
 new Reactor(parentPort);
 
+// This controls how frequently we "release" control from the Atomic; anecdotally
+// this appears to help with stalled wait() on Bun.
+const MAX_WAIT = 500;
+
 class RingBufferReader {
   input: SharedArrayBuffer;
   flag: Int32Array;
   inputOffset: number;
   scratch: ArrayBuffer;
   scratchView: DataView;
-  expected: number;
+  position: number;
+  #available: number;
 
   static SAB_IDX = 0;
-  static SAB_BASE_OFFSET = 4;
 
   constructor(input: SharedArrayBuffer) {
     this.input = input;
-    this.inputOffset = RingBufferReader.SAB_BASE_OFFSET;
+    this.inputOffset = SAB_BASE_OFFSET;
     this.flag = new Int32Array(this.input);
     this.scratch = new ArrayBuffer(8);
     this.scratchView = new DataView(this.scratch);
-    this.expected = 0;
-    this.pull(false);
+    this.position = 0;
+    this.#available = 0;
+    this.wait();
+  }
+
+  close() {
+    this.signal();
+    Atomics.store(this.flag, 0, SAB_BASE_OFFSET);
+  }
+
+  wait() {
+    let value = SAB_BASE_OFFSET;
+    do {
+      value = Atomics.load(this.flag, 0);
+      if (value === SAB_BASE_OFFSET) {
+        const result = Atomics.wait(this.flag, 0, SAB_BASE_OFFSET, MAX_WAIT);
+        if (result === 'timed-out') {
+          continue;
+        }
+      }
+    } while (value <= SAB_BASE_OFFSET);
+
+    this.#available = Atomics.load(this.flag, 0);
```

Comment:
This is the second issue we ran into – sometimes we'd receive a `wait` signal for the flag but, on reading the value, it would turn out to be the old value. The loop ensures we wait until an `Atomics.load` gives us the expected, changed value.